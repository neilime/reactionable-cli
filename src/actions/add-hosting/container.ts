import 'reflect-metadata';

import { Container } from 'inversify';

import Amplify from './adapters/amplify/Amplify';
import { HostingAdapter } from './adapters/HostingAdapter';
import Netlify from './adapters/netlify/Netlify';

export const AdapterKey = 'HostingAdapter';

export function bindHostingAdapters(container: Container): void {
  container.bind<HostingAdapter>(AdapterKey).to(Amplify);
  container.bind<HostingAdapter>(AdapterKey).to(Netlify);
}
