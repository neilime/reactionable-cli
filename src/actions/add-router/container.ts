import "reflect-metadata";

import { Container } from "inversify";

import NextJs from "./adapters/NextJs";
import { RouterAdapter } from "./adapters/RouterAdapter";
import RouterDom from "./adapters/RouterDom";

export const AdapterKey = "RouterAdapter";

export function bindRouterAdapters(container: Container): void {
  container.bind<RouterAdapter>(AdapterKey).to(NextJs);
  container.bind<RouterAdapter>(AdapterKey).to(RouterDom);
}
