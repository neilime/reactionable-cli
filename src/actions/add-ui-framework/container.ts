import 'reflect-metadata';

import { Container } from 'inversify';

import UIBootstrap from './adapters/UIBootstrap';
import { UIFrameworkAdapter } from './adapters/UIFrameworkAdapter';
import UIMaterial from './adapters/UIMaterial';

export const AdapterKey = 'UIFrameworkAdapter';

export function bindUIFrameworkAdapters(container: Container): void {
  container.bind<UIFrameworkAdapter>(AdapterKey).to(UIBootstrap);
  container.bind<UIFrameworkAdapter>(AdapterKey).to(UIMaterial);
}
