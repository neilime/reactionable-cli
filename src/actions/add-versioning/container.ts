import 'reflect-metadata';
import { Container } from 'inversify';
import Git from './adapters/github/Github';
import { IVersioningAdapter } from './IVersioningAdapter';

let container = new Container({ defaultScope: 'Singleton' });

container.bind<IVersioningAdapter>('Adapter').to(Git);

export default container;