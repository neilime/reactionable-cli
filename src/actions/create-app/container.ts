import "reflect-metadata";

import { Container } from "inversify";

import CreateNextApp from "./adapters/create-next-app/CreateNextApp";
import CreateReactApp from "./adapters/create-react-app/CreateReactApp";
import { CreateAppAdapter } from "./adapters/CreateAppAdapter";

export const AdapterKey = "CreateAppAdapter";

export function bindCreateAppAdapters(container: Container): void {
  container.bind<CreateAppAdapter>(AdapterKey).to(CreateReactApp);
  container.bind<CreateReactApp>(CreateReactApp).toSelf();
  container.bind<CreateAppAdapter>(AdapterKey).to(CreateNextApp);
  container.bind<CreateNextApp>(CreateNextApp).toSelf();
}
