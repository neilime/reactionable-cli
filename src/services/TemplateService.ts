import { mkdirSync } from 'fs';
import { resolve, dirname, join, basename, extname, sep } from 'path';
import { plural, singular } from 'pluralize';
import {
  compile,
  registerHelper,
  SafeString,
  registerPartial,
  partials,
} from 'handlebars';
import { inject, injectable } from 'inversify';
import { FileService } from './file/FileService';
import { FileFactory } from './file/FileFactory';
import { StringUtils } from './StringUtils';

// Compile template
registerHelper({
  and: (...parts) => {
    return Array.prototype.slice
      .call(parts, 0, parts.length - 1)
      .every(Boolean);
  },
  eq: (v1, v2) => {
    return v1 === v2;
  },
  neq: (v1, v2) => {
    return v1 !== v2;
  },
  lt: (v1, v2) => {
    return v1 < v2;
  },
  gt: (v1, v2) => {
    return v1 > v2;
  },
  capitalize: (str) => {
    if (typeof str !== 'string') {
      return str;
    }
    return StringUtils.capitalize(str);
  },
  decapitalize: (str) => {
    if (typeof str !== 'string') {
      return str;
    }
    return str.charAt(0).toLowerCase() + str.slice(1);
  },
  pluralize: (str) => {
    if (typeof str !== 'string') {
      return str;
    }
    return plural(str);
  },
  singularize: (str) => {
    if (typeof str !== 'string') {
      return str;
    }
    return singular(str);
  },
  hyphenize: (str) => {
    if (typeof str !== 'string') {
      return str;
    }
    return StringUtils.hyphenize(str);
  },
  camelize: (str) => {
    if (typeof str !== 'string') {
      return str;
    }
    return StringUtils.camelize(str);
  },
  decamelize: (str) => {
    if (typeof str !== 'string') {
      return str;
    }
    return StringUtils.decamelize(str);
  },
  halfSplit: (array) => {
    if (!Array.isArray(array)) {
      return array;
    }
    // Split ranks steps in two parts
    const tmpArray = array;
    const halfWayThough = Math.floor(array.length / 2);
    return [
      tmpArray.slice(0, halfWayThough),
      tmpArray.slice(halfWayThough, tmpArray.length),
    ];
  },
  inline(options) {
    const inline = compile(options.fn(this))(options.data.root);
    const nl2br = inline.replace(/([^>\r\n]?)(\r\n|\n\r|\r|\n)/g, '$1<br>$2');
    return new SafeString(nl2br);
  },
  with(context, options) {
    return options.fn(context);
  },
});

export type TemplateConfig =
  | string[]
  | {
      [key: string]: TemplateConfig | string;
    };

type CompiledTemplate = (context: Object) => string;

@injectable()
export class TemplateService {
  private readonly compiledTemplates: Map<string, CompiledTemplate> = new Map();

  constructor(
    @inject(FileService) private readonly fileService: FileService,
    @inject(FileFactory) private readonly fileFactory: FileFactory
  ) {}

  async renderTemplateTree(
    dirPath: string,
    namespace: string,
    config: TemplateConfig,
    context: Object = {}
  ): Promise<void> {
    this.fileService.assertDirExists(dirPath);

    if (Array.isArray(config)) {
      for (const filePath of config) {
        const currentPath = resolve(
          dirPath,
          await this.renderTemplateString(filePath, context)
        );

        const currentBaseDirPath = dirname(currentPath);
        if (!this.fileService.dirExistsSync(currentBaseDirPath)) {
          mkdirSync(currentBaseDirPath, { recursive: true });
        }

        await this.createFileFromTemplate(currentPath, namespace, context);
      }
      return;
    }

    for (const dir of Object.keys(config)) {
      const templateConfig = config[dir];
      const currentPath = resolve(
        dirPath,
        await this.renderTemplateString(dir, context)
      );

      if (typeof templateConfig === 'string') {
        const currentBaseDirPath = dirname(currentPath);
        if (!this.fileService.dirExistsSync(currentBaseDirPath)) {
          mkdirSync(currentBaseDirPath, { recursive: true });
        }

        await this.createFileFromTemplate(
          currentPath,
          join(namespace, templateConfig),
          context
        );
      } else {
        mkdirSync(currentPath, { recursive: true });
        await this.renderTemplateTree(
          currentPath,
          namespace,
          templateConfig,
          context
        );
      }
    }
  }

  async createFileFromTemplate(
    filePath: string,
    namespace: string,
    context: Object,
    encoding = 'utf8'
  ): Promise<void> {
    const parentDir = dirname(filePath);
    if (!this.fileService.dirExistsSync(parentDir)) {
      throw new Error(
        `Unable to create file "${filePath}", directory "${parentDir}" does not exist`
      );
    }

    const templateKey = extname(namespace)
      ? namespace
      : join(namespace, basename(filePath));
    const fileContent = await this.renderTemplateFile(templateKey, context);

    await this.fileFactory
      .fromString(fileContent, filePath, encoding)
      .saveFile();
  }

  async getTemplateFileContent(template: string): Promise<string> {
    const templatePath = join('./../templates', template + '.template');

    let templateContent: string;
    try {
      const importedContent = await import(templatePath);
      if ('string' === typeof importedContent) {
        templateContent = importedContent;
      } else if (
        'object' === typeof importedContent &&
        'string' === typeof importedContent.default
      ) {
        templateContent = importedContent.default;
      } else {
        throw new Error(
          `Unexpected content retrieved from importing template file "${templatePath}": ${typeof importedContent}`
        );
      }
    } catch (error) {
      throw new Error(
        `An error occurred while importing template file "${templatePath}"`
      );
    }

    // Register partials if any
    const regex = /\{\{#> ([a-zA-Z]+) \}\}/gim;
    let matches;
    while ((matches = regex.exec(templateContent)) !== null) {
      if (matches.index === regex.lastIndex) {
        regex.lastIndex++;
      }
      const partialName: string = matches[1];

      if (partials[partialName]) {
        continue;
      }

      const partialTemplateKey = `${
        template.split(sep)[0]
      }/partials/${partialName}`;
      const partialTmplateContent = await this.getTemplateFileContent(
        partialTemplateKey
      );

      registerPartial(partialName, partialTmplateContent);
    }

    return templateContent;
  }

  async getCompiledTemplateString(
    templateKey: string,
    templateContent: string
  ): Promise<CompiledTemplate> {
    let compiledTemplate = this.compiledTemplates.get(templateKey);
    if (compiledTemplate) {
      return compiledTemplate;
    }
    compiledTemplate = compile(templateContent);

    this.compiledTemplates.set(templateKey, compiledTemplate);

    return compiledTemplate;
  }

  async renderTemplateString(
    template: string,
    context: Object
  ): Promise<string> {
    const compiledTemplate = await this.getCompiledTemplateString(
      template,
      template
    );
    return compiledTemplate(context);
  }

  async getCompiledTemplateFile(
    templateKey: string
  ): Promise<CompiledTemplate> {
    let compiledTemplate = this.compiledTemplates.get(templateKey);
    if (compiledTemplate) {
      return compiledTemplate;
    }

    const templateFileContent = await this.getTemplateFileContent(templateKey);
    try {
      compiledTemplate = compile(templateFileContent);

      this.compiledTemplates.set(templateKey, compiledTemplate);

      return compiledTemplate;
    } catch (error) {
      throw new Error(
        `An error occurred while compiling template "${templateKey}": ${error}`
      );
    }
  }

  async renderTemplateFile(
    templateKey: string,
    context: Object
  ): Promise<string> {
    const compiledTemplate = await this.getCompiledTemplateFile(templateKey);
    try {
      const content = compiledTemplate(context);
      return content;
    } catch (error) {
      throw new Error(
        `An error occurred while rendering template "${templateKey}": ${error}`
      );
    }
  }
}
