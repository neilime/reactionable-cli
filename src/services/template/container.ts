import "reflect-metadata";

import { Container } from "inversify";

import { EtaAdapter } from "./adapters/EtaAdapter";
import { TemplateAdapterHelper } from "./adapters/TemplateAdapterHelper";

export const AdapterKey = "TemplateAdapter";

export function bindTemplateAdapters(container: Container): void {
  container.bind<TemplateAdapterHelper>(TemplateAdapterHelper).toSelf();
  container.bind<EtaAdapter>(AdapterKey).to(EtaAdapter);
  container.bind<EtaAdapter>(EtaAdapter).toSelf();
}
