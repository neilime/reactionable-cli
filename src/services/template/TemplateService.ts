import { dirname, extname, join, resolve } from "path";

import { inject, injectable } from "inversify";

import { FileFactory } from "../file/FileFactory";
import { FileService } from "../file/FileService";
import { TemplateAdapter } from "./adapters/TemplateAdapter";
import { AdapterKey } from "./container";
import { TemplateContext } from "./TemplateContext";
import { TemplateFileService } from "./TemplateFileService";

type TemplateConfigItem = string[] | TemplateConfig | string;

export type TemplateConfig = {
  [key: string]: TemplateConfigItem;
};

@injectable()
export class TemplateService {
  constructor(
    @inject(FileService) private readonly fileService: FileService,
    @inject(FileFactory) private readonly fileFactory: FileFactory,
    @inject(TemplateFileService) private readonly templateFileService: TemplateFileService,
    @inject(AdapterKey) private readonly templateAdapter: TemplateAdapter
  ) {}

  async renderTemplate(
    dirPath: string,
    namespace: string,
    context: TemplateContext = {}
  ): Promise<void> {
    this.fileService.assertDirExists(dirPath);

    const config: TemplateConfig = await this.getTemplateConfig(namespace, context);
    return this.renderTemplateFromConfig(dirPath, namespace, context, config);
  }

  async renderTemplateFile(templateKey: string, context: TemplateContext): Promise<string> {
    return this.templateAdapter.renderTemplateFile(templateKey, context);
  }

  /**
   * Render the template file(s) form given config
   */
  private async renderTemplateFromConfig(
    dirPath: string,
    namespace: string,
    context: TemplateContext,
    config: TemplateConfigItem
  ): Promise<void> {
    if (Array.isArray(config)) {
      for (const filePath of config) {
        await this.renderTemplateFromConfig(dirPath, namespace, context, filePath);
      }
      return;
    }

    if (typeof config === "object") {
      for (const dir of Object.keys(config)) {
        const currentPath = resolve(dirPath, dir);
        await this.renderTemplateFromConfig(currentPath, namespace, context, config[dir]);
      }
      return;
    }

    const currentNamespace = join(namespace, config);
    const currentPath = extname(dirPath) ? dirPath : join(dirPath, config);
    const currentBaseDirPath = dirname(currentPath);

    if (!this.fileService.dirExistsSync(currentBaseDirPath)) {
      this.fileService.mkdirSync(currentBaseDirPath, true);
    }

    const templateKey = this.getTemplateKey(dirPath, currentPath, currentNamespace);

    await this.createFileFromTemplate(currentPath, templateKey, context);
  }

  async getTemplateConfig(namespace: string, context: TemplateContext): Promise<TemplateConfig> {
    const templateKey = join(namespace, "config");
    const templateContent = await this.templateFileService.getTemplateFileContent(templateKey);
    const content = await this.templateAdapter.renderTemplateString(templateContent, context);
    const templateConfig = JSON.parse(content);
    return templateConfig;
  }

  getTemplateKey(dirPath: string, filepath: string, namespace: string): string {
    // Namespace is the template key
    if (extname(namespace)) {
      return namespace;
    }

    const templateKey = join(namespace, filepath.replace(dirPath, ""));
    const templatePath = join(__dirname, "./../templates", templateKey + ".template");
    if (!this.fileService.fileExistsSync(templatePath)) {
      throw new Error(`Template file "${templatePath}" does not exist`);
    }

    return templateKey;
  }

  async createFileFromTemplate(
    filePath: string,
    templateKey: string,
    context: TemplateContext,
    encoding: BufferEncoding = "utf8"
  ): Promise<void> {
    const parentDir = dirname(filePath);
    if (!this.fileService.dirExistsSync(parentDir)) {
      throw new Error(
        `Unable to create file "${filePath}", directory "${parentDir}" does not exist`
      );
    }

    const fileContent = await this.templateAdapter.renderTemplateFile(templateKey, context);

    await this.fileFactory.fromString(fileContent, filePath, encoding).saveFile();
  }
}
