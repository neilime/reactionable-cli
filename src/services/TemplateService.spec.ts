import container from '../container';
import { TemplateService } from './TemplateService';

describe('templateService', () => {
  let service: TemplateService;

  beforeEach(() => {
    // Initialize service before each test to not be confused by cache
    container.snapshot();
    service = container.get(TemplateService);
  });

  afterEach(() => {
    container.restore();
  });

  describe('createFileFromTemplate', () => {
    it('should throws an error if file directory does not exist', async () => {
      const createFileOperation = service.createFileFromTemplate(
        '/unexisting-directory/test.js',
        'test-namespace',
        {}
      );

      await expect(createFileOperation).rejects.toThrow(
        `Unable to create file "/unexisting-directory/test.js", directory "/unexisting-directory" does not exist`
      );
    });
  });
});
