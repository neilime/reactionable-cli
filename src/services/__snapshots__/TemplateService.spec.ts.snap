// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`templateService renderTemplate should create a file from given namespace 1`] = `
"import { initializeI18n } from '';
import enCommon from './locales/en/common.json';
import frCommon from './locales/fr/common.json';

const resources = {
  en: { 
    common: enCommon,
  },
  fr: { 
    common: frCommon,
  },
};


initializeI18n({ resources });"
`;

exports[`templateService renderTemplate should create a file from given namespace 2`] = `
"{
  \\"test-project\\": \\"test-project\\"
}"
`;

exports[`templateService renderTemplate should create a file from given namespace having nested config: TestEntitiesConfig.tsx 1`] = `
"import { FormField } from '';
import { IFormChildrenProps } from '@reactionable/core';
import { TFunction } from 'i18next';
import React from 'react';
import { string } from 'yup';

export type ITestEntityData = {
  id: string;
  label: string;
};

export type ITestEntityValues = Omit<ITestEntityData, \\"id\\">;


export type IGetTestEntityQueryVariables = Pick<ITestEntityData, \\"id\\"> & {

};


export type IListTestEntitiesQueryVariables = {
  
};


export const testEntityFormValues: ITestEntityValues = {
  label: \\"\\",
};


export const testEntityFormSchema = (t: TFunction) => ({
   label: string()
    .min(1, t(\\"Test entity label must be at least 1 characters\\"), { ns : \\"testEntities\\" })
    .required(t(\\"Test entity label is required\\"), { ns : \\"testEntities\\" }),
});


export const testEntityFormChildren = (t: TFunction) => (formikProps: IFormChildrenProps<ITestEntityValues>) => <>
  <FormField
    name=\\"label\\"
    label={t(\\"Label\\", { ns : \\"testEntities\\" })}
    placeholder={t(\\"Label\\", { ns : \\"testEntities\\" })}
    autoFocus={true}
  />
</>;"
`;

exports[`templateService renderTemplate should create a file from given namespace having nested config: testEntities.json 1`] = `
"{
  \\"Test entity label must be at least 1 characters\\": \\"Test entity label must be at least 1 characters\\",
  \\"Test entity label is required\\": \\"Test entity label is required\\",
  \\"Label\\": \\"Label\\",
  \\"Create a new test entity\\": \\"Create a new test entity\\",
  \\"Delete test entity\\": \\"Delete test entity\\",
  \\"Are you sure you want to delete this test entity ?\\": \\"Are you sure you want to delete this test entity ?\\",
  \\"The test entity has been deleted\\": \\"The test entity has been deleted\\",
  \\"Go to test entity \\\\\\"{{label}}\\\\\\"\\": \\"Go to test entity \\\\\\"\\\\\\"\\",
  \\"Delete test entity \\\\\\"{{label}}\\\\\\"\\": \\"Delete test entity \\\\\\"\\\\\\"\\",
  \\"You don\\\\\\"t have any test entities yet\\": \\"You don\\\\\\"t have any test entities yet\\",
  \\"Click the button to create a new test entity\\": \\"Click the button to create a new test entity\\",
  \\"Test entity - {{label}}\\": \\"Test entity - {{label}}\\",
  \\"Update test entity \\\\\\"{{label}}\\\\\\"\\": \\"Update test entity \\\\\\"{{label}}\\\\\\"\\"
}"
`;
